### PR TITLE
Add deprecation policy to plugin writer guide

### PR DESCRIPTION
**Pull request by bmbouter**
* Adds docs on how the deprecation process works
* Adds docs on how to depend on pulpcore with this policy

closes #7413

### CODE REVIEW

File: docs/plugins/plugin-writer/concepts/index.rst [hunk 1]
```diff
@@ -328,6 +328,45 @@ For the MANIFEST.in entry, you'll likely want one like the example below which w
...
+
+Then in 3.9 the following happens:
+
+1. The ``def foo(a, b)`` method is deleted entirely.
+2. A ``CHANGES/plugin_api/XXXX.removal`` changelog entry is created explaining what has been
+   removed.
+
+.. note::
+
+    Python ``DeprecationWarning`` log statements are show to users of your plugin when using a
```
Comment (goosemania) [inline docs/plugins/plugin-writer/concepts/index.rst:361]: s/show/shown/

File: docs/plugins/reference/releasing.rst [hunk 2]
```diff
@@ -4,12 +4,20 @@ Releasing Your Plugin
...
+it safe for a plugin compatible with, e.g. pulpcore 3.7 to declare compatibility up to pulpcore 3.8.
+In this example, use the following requirements string::
 
-    pulpcore>=3.0,<3.2
+    pulpcore>=3.7,<3.9
 
-This ensures that if pulpcore==3.2 releases and contains backwards incompatible changes, your users
-won't upgrade into an environment with an incompatible pulpcore.
+This ensures that when pulpcore 3.8 is released, your users can receive it immediately and use it
+without issue. However when 3.9 comes out, any deprecations introduced in the ``pulpcore.plugin``
```
Comment (goosemania) [inline docs/plugins/reference/releasing.rst:14]: maybe without issues? or without any issue?
I thought that "without issue" often has an idiomatic use but I can easily be wrong.
Ignore me, if it's ok. Just wanted to bring your attention to check.

File: docs/plugins/reference/releasing.rst [hunk 3]
```diff
@@ -4,12 +4,20 @@ Releasing Your Plugin
...
+API in 3.8 will be removed, so it preventing your plugin from working with pulpcore 3.9 maintains
```
Comment (goosemania) [inline docs/plugins/reference/releasing.rst:15]: s/it//?

File: docs/plugins/reference/releasing.rst [hunk 4]
```diff
@@ -4,12 +4,20 @@ Releasing Your Plugin
 Depending on pulpcore
 ---------------------
 
-A plugin writer should specify specify the minimum and maximum version of pulpcore their plugin
-requires. The Plugin API is not yet stable so plugins should pin on a specific 3.y version of
-pulpcore. For example, if your plugin requires 3.0 and is also compatible with pulpcore 3.1, you
-want to restrict it to::
+The Plugin API is not yet stable, but has a :ref:`deprecation process <deprecation_policy>`, making
```
Comment (goosemania) [inline docs/plugins/reference/releasing.rst:7]: Does it make sense to note somewhere that we have deprecation policy starting with 3.7?
Up to you.

**Review by goosemania — APPROVED** [review]
Thanks!

I left just some minor comments, no need to re-review.

File: docs/plugins/plugin-writer/concepts/index.rst [hunk 5]
```diff
@@ -328,6 +328,45 @@ For the MANIFEST.in entry, you'll likely want one like the example below which w
...
 
+.. _deprecation_policy:
+
+Plugin API Stability and Deprecation Policy
+-------------------------------------------
+
+The ``pulpcore.plugin`` API can introduce breaking changes, and will be introduced in the following
+way. For this example, assume that pulpcore 3.8 introduces a breaking change by changing the call
+signature of a method named ``def foo(a, b)`` which is importable via the plugin API. Here are the
+changes that would happen:
```
Comment (ipanova) [inline docs/plugins/plugin-writer/concepts/index.rst:339]: this sentence is redundant

File: docs/plugins/plugin-writer/concepts/index.rst [hunk 6]
```diff
@@ -328,6 +328,45 @@ For the MANIFEST.in entry, you'll likely want one like the example below which w
...
+In 3.8 the following changes happen:
+
+1. The new method would be introduced as a new named function ``def the_new_foo(...)`` or some
+   similar name.
+2. The existing method signature ``def foo(a, b)`` is left in-tact.
+3. The ``foo`` method would have the a Python ``DeprecationWarning`` added to it such as::
+
+    warnings.warn("foo() is deprecated and will be removed in pulpcore==3.9; use the_new_foo().", warnings.DeprecationWarning)
+
+4. A ``CHANGES/plugin_api/XXXX.deprecated`` changelog entry is created explaining how to port plugin
```
Comment (daviddavis) [inline docs/plugins/plugin-writer/concepts/index.rst:350]:
```suggestion
4. A ``CHANGES/plugin_api/XXXX.deprecation`` changelog entry is created explaining how to port plugin
```

File: docs/plugins/reference/releasing.rst [hunk 7]
```diff
@@ -4,12 +4,20 @@ Releasing Your Plugin
...
+it safe for a plugin compatible with, e.g. pulpcore 3.7 to declare compatibility up to pulpcore 3.8.
+In this example, use the following requirements string::
 
-    pulpcore>=3.0,<3.2
+    pulpcore>=3.7,<3.9
 
-This ensures that if pulpcore==3.2 releases and contains backwards incompatible changes, your users
-won't upgrade into an environment with an incompatible pulpcore.
+This ensures that when pulpcore 3.8 is released, your users can receive it immediately and use it
```
Comment (ipanova) [inline docs/plugins/reference/releasing.rst:13]: s/your users/users

File: docs/plugins/reference/releasing.rst [hunk 8]
```diff
@@ -4,12 +4,20 @@ Releasing Your Plugin
...
+without issue. However when 3.9 comes out, any deprecations introduced in the ``pulpcore.plugin``
+API in 3.8 will be removed, so it preventing your plugin from working with pulpcore 3.9 maintains
+compatability.
```
Comment (ipanova) [inline docs/plugins/reference/releasing.rst:16]: compatibility

File: docs/plugins/reference/releasing.rst (hunk 2 above)
Comment (bmbouter) [reply docs/plugins/reference/releasing.rst:14]: I switched to `without any issue`.

File: docs/plugins/plugin-writer/concepts/index.rst (hunk 5 above)
Comment (bmbouter) [reply docs/plugins/plugin-writer/concepts/index.rst:339]: removed, ty

**Review by daviddavis — APPROVED** [review]

**Pull request merged**